libdrgn/dwarf_info_cache.c

```diff
@@ -1475,6 +1485,52 @@ drgn_dwarf_object_find(const char *name, size_t name_len, const char *filename,
 	return &drgn_not_found;
 }
 
+struct drgn_error *
+drgn_dwarf_object_find_with_namespaces(const char *name, size_t name_len, const char *filename,
+				       enum drgn_find_object_flags flags,
+				       struct drgn_dwarf_index_namespace* ns,
+				       void *arg, struct drgn_object *ret) {
+	struct drgn_error *err;
+	struct drgn_dwarf_info_cache *dicache = arg;
+	char* ns_ptr;
+
+	if (!ns)
+		ns = &dicache->dindex.ns;
+
+find_ns:
+	ns_ptr = strstr(name, "::");
+	if (ns_ptr) {
+		const int ns_len = ns_ptr - name;
+		const size_t ns_colon_len = ns_len + 2;
+		struct drgn_dwarf_index_iterator it;
+		uint64_t ns_tag = DW_TAG_namespace;
+		struct drgn_dwarf_index_die *die;
+
+		err = drgn_dwarf_index_iterator_init(&it, ns, name, ns_len, &ns_tag, 1, true);
+		if (err)
+			return err;
+		// Namespaces should be de-duplicated so we only check for the first one.
+		while ((die = drgn_dwarf_index_iterator_next_die(&it))) {
+			assert(die->tag == DW_TAG_namespace);
+			name = name + ns_colon_len;
+			ns = die->namespace;
+			// Dive into this namespace
+			goto find_ns;
+
+		}
+		// Fall-through to default implementation
+	}
+
+	return drgn_dwarf_object_find_internal(name, name_len, filename, flags, ns, arg, ret);
+}
+
+struct drgn_error *
+drgn_dwarf_object_find(const char *name, size_t name_len, const char *filename,
+		       enum drgn_find_object_flags flags, void *arg,
+		       struct drgn_object *ret) {
+	return drgn_dwarf_object_find_with_namespaces(name, name_len, filename, flags, NULL, arg, ret);
```

Comment:
I'd just combine `drgn_dwarf_object_find()`, `drgn_dwarf_object_find_with_namespaces()`, and `drgn_dwarf_object_find_internal()`.